modulus/distributed/utils.py

```diff
@@ -218,3 +219,293 @@ def _gather(input_, dim_, group=None):
+
+    comm_size = dist.get_world_size(group=group)
+    rank = dist.get_rank(group=group)
+    assert len(sizes) == comm_size
+    assert dim < tensor.dim()
+
+    global_size = sum(sizes)
+
+    if comm_size == 1:
+        return tensor
+
+    tensor_shape = list(tensor.shape)
+    tensor_list = [None] * comm_size
+
+    for src in range(comm_size):
+        tensor_shape[dim] = sizes[src]
+        tensor_list[src] = torch.empty(
+            tensor_shape,
+            dtype=tensor.dtype,
+            device=tensor.device,
+        )
+
+    dist.all_gather(tensor_list, tensor, group=group)
+
+    output = torch.cat(tensor_list, dim=dim)
+
+    return output
+
+
+def all_reduce_v_wrapper(
+    tensor: torch.Tensor,
+    sizes: List[int],
+    dim: int = 0,
+    use_fp32: bool = True,
+    group: Optional[dist.ProcessGroup] = None,
+) -> torch.Tensor:
+    """implementation of a distributed all_reduce_v primitive"""
+
+    comm_size = dist.get_world_size(group=group)
+    rank = dist.get_rank(group=group)
+    assert len(sizes) == comm_size
+    assert dim < tensor.dim()
+
+    global_size = sum(sizes)
+
+    tensor_shape = list(tensor.shape)
+    tensor_shape[dim] = sizes[rank]
+    tmp = [
+        torch.empty(
+            tensor_shape,
+            dtype=tensor.dtype,
+            device=tensor.device,
+        )
+        for _ in range(comm_size)
+    ]
+    scatter_list = list(torch.split(tensor, sizes, dim=dim))
+
+    dist.all_to_all(tmp, scatter_list, group=group)
+    stack_dim = tensor.dim()
+    tmp = torch.stack(tmp, dim=stack_dim)
+
+    if use_fp32:
+        # cast to float before sum and return float, then cast back
+        output = tmp.sum(dim=stack_dim, dtype=torch.float32)
+        output = output.to(dtype=tensor.dtype)
+    else:
+        # else: just do sum in native dtype
+        output = tmp.sum(dim=stack_dim)
+
+    return output
+
+
+def gather_v_wrapper(
+    tensor: torch.Tensor,
+    sizes: List[int],
+    dim: int = 0,
+    dst: int = 0,
+    group: Optional[dist.ProcessGroup] = None,
+) -> torch.Tensor:
+    """implementation of a distributed gather_v primitive"""
+
+    comm_size = dist.get_world_size(group=group)
+    rank = dist.get_rank(group=group)
+    assert len(sizes) == comm_size
+    assert 0 <= dst < comm_size
+    assert dim < tensor.dim()
+
+    global_size = sum(sizes)
+
+    if comm_size == 1:
+        return tensor
+
+    tensor_shape = list(tensor.shape)
+    tensor_list = [None] * comm_size
+
+    if rank == dst:
+        for r in range(comm_size):
+            req_list = [None] * comm_size
+            if r == dst:
+                tensor_list[dst] = tensor
+            else:
+                tensor_shape[dim] = sizes[r]
+                tensor_list[r] = torch.empty(
+                    tensor_shape,
+                    device=tensor.device,
+                    dtype=tensor.dtype,
+                )
+
+                req_list[r] = dist.irecv(tensor_list[r], src=r, group=group)
```

Comment:
Why not just use `torch.distributed.gather` here? From the documentation, seems like it supports variable tensor sizes across processes: https://pytorch.org/docs/stable/distributed.html#torch.distributed.gather

Reply:
Actually this is a good point. In a few places, the documentation clearly states that tensors should have the same size while here it only specifies `appropriately-sized tensors`. I am not sure if I tried gather but I thought that I indeed saw issues when tensors don't have the same size using just plain gather.